### PR TITLE
Calculate_think_time

### LINKED ISSUE #73: Task: Write Trails Time Feature functions: think_time
### Description

Calculate total and mean think times using consecutive segments approach. Think time at a circle = time from entering the circle (end of incoming segment) to leaving the circle (start of outgoing segment). This function will be based off the calculate_think_times function from https://github.com/theBasicScientist/trailsab-proc/blob/main/trails_analyzer.py.

### Tasks

- [ ] Sort segments to ensure proper order (by their actual sequence in the task)
- [ ] Find consecutive segments: The destination of current segment should equal origin of next segment
- [ ] Find when we ENTERED this circle (working backwards from end of current segment)
- [ ] Find when we LEFT this circle (working forwards from start of next segment)
- [ ] Calculate think time and assign to final list[think_times]
- [ ] Calculate sum and mean of list[think_times]

### Freeform Notes

_No response_

### PR DESCRIPTION
**Pull request by cgmaiorano**
Resolves #73 

Broke up longer function from https://github.com/theBasicScientist/trailsab-proc/blob/main/trails_analyzer.py#L86 into private functions for calculating circle entry time and exit time. Tests behavior for calculating intermediate think times and separate tests for edge cases.

### CODE REVIEW

File: tests/unit/test_trails_time.py [hunk 1]
```diff
@@ -71,3 +74,187 @@ def test_error_at_start() -> None:
...
+                    "C",
+                    [
+                        {"x": 5.0, "y": 5.0, "seconds": 3.0},
+                        {"x": 20.0, "y": 20.0, "seconds": 5.0},
+                    ],
+                ),
+            ],
+            {"B": ("B", 5.0, 5.0, 1.0)},
+            [{"label": "A", "order": 1}, {"label": "B", "order": 2}],
+            [("B", "C", 3.0, "B")],
```
Comment (kimit0310) [inline tests/unit/test_trails_time.py]: Could we make these expected values have some description to them by either adding a description field or like an inline comment? the id does explain the scenario, but it doesn't explain how the 3.0 was derived.

File: tests/unit/test_trails_time.py [hunk 2]
```diff
@@ -71,3 +74,187 @@ def test_error_at_start() -> None:
...
+            id="no_think_time_when_circle_label_missing",
+        ),
+        pytest.param(
+            [
+                ("A", "B", [{"x": 5.0, "y": 5.0, "seconds": 5.0}]),
+                ("B", "C", [{"x": 5.0, "y": 5.0, "seconds": 3.0}]),
+            ],
+            {"B": ("B", 5.0, 5.0, 1.0)},
+            [{"label": "A", "order": 1}, {"label": "B", "order": 2}],
+            [],
```
Comment (kimit0310) [inline tests/unit/test_trails_time.py]: pretty much the same comment here as line 153, could we include a description field in the test parameters that explains why no think time is expected?

File: tests/unit/test_trails_time.py [hunk 3]
```diff
@@ -71,3 +74,187 @@ def test_error_at_start() -> None:
...
+
+    expected_by_key = {(s, e): (tt, lbl) for s, e, tt, lbl in expected}
+    for seg in result:
+        key = (seg.start_label, seg.end_label)
+        if key in expected_by_key:
+            think_time, circle_label = expected_by_key[key]
+            assert seg.think_time == think_time
+            assert seg.think_circle_label == circle_label
+        else:
+            assert seg.think_time == 0.0
```
Comment (kimit0310) [inline tests/unit/test_trails_time.py]: shouldn't we also check the default `think_circle_label` which I guess would be an `""`

File: tests/unit/test_trails_time.py [hunk 4]
```diff
@@ -71,3 +74,187 @@ def test_error_at_start() -> None:
...
+    config: Dict[str, Dict[str, List[Dict[str, Any]]]] = {
+        "trail1": {"items": [{"label": "A", "order": 1}, {"label": "B", "order": 2}]}
+    }
+
+    result = time.calculate_think_times(segments, circles, config, "trail1")
+
+    assert [seg.start_label for seg in result] == ["A", "B"]
+
+
+def test_empty_segments_returns_empty() -> None:
```
Comment (kimit0310) [inline tests/unit/test_trails_time.py]: How about when `segments[0].points` is empty? we have a guard for it in line 129 on `time.py`, I feel like we would need a test for that too.

File: src/graphomotor/features/trails/time.py [hunk 5]
```diff
@@ -1,5 +1,9 @@
 """Feature extraction module for time-based metrics in trails drawing data."""
 
+from typing import Any, Dict, List, Optional
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: Very nitty gritty but `calculate_total_error_time` uses `dict[str, float]` for type hints, so we could keep constistancy.

File: src/graphomotor/features/trails/time.py [hunk 6]
```diff
@@ -50,3 +54,90 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
         total_error_time += end_time - start_time
 
     return {"total_error_time": float(total_error_time)}
+
+
+def _find_circle_entry_time(
+    points: pd.DataFrame, circle: models.CircleTarget
+) -> Optional[float]:
+    """Find when a point last entered the circle by scanning backwards."""
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: I think we could use a little more detail to this docstring. This returns the seconds value from the last sampled point that's inside the circle, which is kind of different from calculating or interpolating the boundary-crossing time, which I think this could be interpreted as. We could also just change the function name to something like `find_last_time_inside_circle` or something like that

File: src/graphomotor/features/trails/time.py [hunk 7]
```diff
@@ -50,3 +54,90 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
...
+            if circle.contains_point(points.iloc[i]["x"], points.iloc[i]["y"])
+        ),
+        None,
+    )
+
+
+def _find_circle_exit_time(
+    points: pd.DataFrame, circle: models.CircleTarget
+) -> Optional[float]:
+    """Find when a point first left the circle by scanning forward."""
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: Same comment as above (line 62), but I also think that we should explain the fallback behavior in the docstring as well.

File: src/graphomotor/features/trails/time.py [hunk 8]
```diff
@@ -50,3 +54,90 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
...
+        config: Configuration dictionary containing circle order for trails (output
+            of create_config_from_circles)
+        trail_id: Trail identifier for circle lookup
+
+    Returns:
+        Updated list of segments with think times calculated
+    """
+    trail_circles = circles[trail_id]
+    circle_order = {item["label"]: item["order"] for item in config[trail_id]["items"]}
+    segments = sorted(segments, key=lambda s: circle_order.get(s.start_label, 999))
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: Could we use `float("inf")` instead of 999 as sort key fallback? or maybe keep it and put it in the docstring with a logger warning?

File: src/graphomotor/features/trails/time.py [hunk 9]
```diff
@@ -50,3 +54,90 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
...
+        next_points = next_seg.points.reset_index(drop=True)
+
+        entry_time = _find_circle_entry_time(current_points, circle)
+        exit_time = _find_circle_exit_time(next_points, circle)
+
+        if entry_time is not None and exit_time is not None and exit_time > entry_time:
+            next_seg.think_time = exit_time - entry_time
+            next_seg.think_circle_label = circle_label
+
+    if not segments or len(segments[0].points) == 0:
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: Could we turn these nested ifs into early return guards?

**Review by kimit0310 — COMMENTED** [review]
Just a few small things + the nested ifs vs early return guards thing.

File: src/graphomotor/features/trails/time.py [hunk 10]
```diff
@@ -56,88 +58,120 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
...
-    segments = sorted(segments, key=lambda s: circle_order.get(s.start_label, 999))
+    target_circles = circle_mapping[trail_id]
+
+    first_seg = segments[0]
+    if first_seg.start_label in target_circles:
+        exit_time = _find_circle_exit_time(
+            first_seg.points, target_circles[first_seg.start_label]
+        )
+
+        if exit_time is not None:
```
Comment (kimit0310) [inline src/graphomotor/features/trails/time.py]: I think we should do inverse so that if exit_time is None throw warning

File: src/graphomotor/features/trails/time.py [hunk 11]
```diff
@@ -50,3 +56,122 @@ def calculate_total_error_time(drawing: models.Drawing) -> dict[str, float]:
...
+    This function computes all the think_times for all provided LineSegments.
+    `think_time` is defined as the difference in timestamps between the first point of
+    the current LineSegment entering a circle and the first point of the next
+    LineSegment exiting that same circle.
+
+    For the first segment, if it starts inside a circle, the `think_time` is calculated
+    as the difference between the first point of the segment and the first point exiting
+    that circle.
+
+    This method is only for segments without errors.
```
Comment (Asanto32) [inline src/graphomotor/features/trails/time.py]: improve to include our new definition of think time that is defined only on correct segments, thus exit time includes the erroneous adventures,
and because only correct segments the start and end MUST be in circles

File: src/graphomotor/features/trails/time.py (hunk 10 above)
Comment (Asanto32) [reply src/graphomotor/features/trails/time.py]: removed, covered by using only correct LineSegments

File: tests/unit/test_trails_time.py [hunk 12]
```diff
@@ -221,9 +224,89 @@ def test_calculate_think_times(
 
     time.calculate_think_times(segments, circle_mapping, "5555555")
 
-    assert segments[0].think_time == 1.0
```
Comment (kimit0310) [inline tests/unit/test_trails_time.py:224]: I think these should be pytest.approx(1.5) instead since they are floats
